DS Scheduler changes

trshaffer opened the pull request:
Here's a preliminary PR with my DS changes. The highlights:

### Refactored the RPC layer

Before we were building and unpacking JSON structures everywhere and had repeated code to deal with the RPC envelopes. I added helper functions for generating, validating, and unpacking the RPC part so we don't need to repeat that stuff. The new functions look like:

    if(ds_unpack_request(msg, &method, &id, &params) != DS_RESULT_SUCCESS) {
        ...

    struct jx *msg = ds_message_request("blob-commit", jx_objectv("blob-id", jx_string(blobid), NULL));

### Switch to JSON-RPC

We had gotten away from JSON-RPC, and I think it's better if we avoid another internal scheme that needs to be documented and taught to anyone making a client library (we already have MQ's wire format....). So the helper functions operate on compliant JSON-RPC messages. We weren't getting any additional functionality by using a custom RPC format, it was just different.

### Change client and worker tables to sets

We were using heap allocated stringified pointers to identify client/worker reps in the manager, which was painful to work with. This PR switches to `struct set`s of just the pointers. Note that this depends on #2475 somewhat, (if workers are persistent, we might need UUIDs and a hash table of those).

dthain commented:
Ok, I like the direction this is going.  If it compiles, please merge it so the rest of us can build on it.

trshaffer commented:
This really turned into a big one, sorry for blocking things for so long.... But the good news is, I can now successfully run a task all the way through from client to worker and back! Additional changes from last time:

### Fix signal handling for child processes
To be able to break out of MQ wait promptly when processes exit,`SIGCHLD` needs to be blocked at the start of the program and a no-op signal handler needs to be defined. Then MQ unblocks only while waiting. Note that there's still a race condition on platforms that don't support that as an atomic operation.

### Switch the manager to using `ds_rpc.c`
Removes the (mostly no-op) handlers in `ds_manager.c` and relies on the RPC layer to handle worker messages.

### Wire in task scheduler
The dummy task scheduler (choose worker at random) is wired in and working. I didn't do blob scheduling yet, though.

### Implement client wait
Tasks now remember who created them, so on a state change we can send notifications to interested clients. Clients will receive a list of updates on calling `wait()` (or block until an update becomes available). I didn't implement wait timeouts yet.

I also added the first client handshake option: `nowait`, since it was only a few lines of code and would be useful for clients that support async IO. See the commit message for details.

### Rework task structs
Ben had already split the states for tasks vs. attempts. I rearranged the structs too, so rather than `struct task` serving both roles and having a `struct task_rep` only on the manager, it's now
- `struct task` gives the static definition of a task, and on the manager we also store some extra info like which worker it's scheduled to. Also holds a list of attempts.
- `struct task_attempt` stores the detailed results of executing a task in a process/on a worker. This would be a good place to store extra info like exit status, resource utilization, etc. Has some extra fields used on the manger for remembering what the worker is up to.

## Running a task
There's enough implemented to run a task from beginning to end. I only got a small set of RPCs working:
- client task submit
- worker task submit
- worker task update
- client wait

The rest of the API still needs doing of course, but the core machinery is put together. I also added a tiny dummy app that executes a single command to test this out. Note that the output below is interleaved. I started the manager, then the worker, then the client.

Note that client recovery from disk is segfaulting, I didn't really look into that. I've just been using a fresh client state.

On the manger:

    $ ./ds_manager -dall -p1234
    tcp: listening on port 1234
    dataswarm: listening on port 1234...
    dataswarm: advertising to the catalog server(s) at catalog.cse.nd.edu,backup-catalog.cse.nd.edu ...
    dns: looking up name catalog.cse.nd.edu
    dns: catalog.cse.nd.edu is 129.74.152.177
    debug: sending update via udp to catalog.cse.nd.edu(129.74.152.177):9097
    dns: looking up name backup-catalog.cse.nd.edu
    dns: backup-catalog.cse.nd.edu is 129.74.247.43
    debug: sending update via udp to backup-catalog.cse.nd.edu(129.74.247.43):9097
    tcp: got connection from 127.0.0.1 port 45594
    dataswarm: new connection from 127.0.0.1:45594
    dataswarm: rx: {"params":{"type":"worker"},"method":"handshake","jsonrpc":"2.0"}
    dataswarm: new worker from 127.0.0.1:45594
    dataswarm: rx: {"params":{"hello":"manager"},"method":"status-report","jsonrpc":"2.0"}
    dataswarm: worker 127.0.0.1:45594 rx: status-report
    tcp: got connection from 127.0.0.1 port 45598
    dataswarm: new connection from 127.0.0.1:45598
    dataswarm: rx: {"method": "handshake", "params": {"type": "client"}, "jsonrpc": "2.0"}
    dataswarm: new client from 127.0.0.1:45598
    dataswarm: rx: {"method": "task-submit", "params": {"command": "exit 123"}, "id": 1, "jsonrpc": "2.0"}
    dataswarm: json tx: {"result":"A6AEF6F6-1C19-4DC9-9C0F-C88BC77A1F40","id":1,"jsonrpc":"2.0"}
    dataswarm: json tx: {"id":0,"params":{"state":"active","resources":{},"task-id":"A6AEF6F6-1C19-4DC9-9C0F-C88BC77A1F40","command":"exit 123"},"method":"task-submit","jsonrpc":"2.0"}
    dataswarm: rx: {"method": "wait", "params": {"timeout": 0}, "id": 2, "jsonrpc": "2.0"}
    dataswarm: rx: {"result":0,"id":0,"jsonrpc":"2.0"}
    dataswarm: worker 127.0.0.1:45594 rx: ok (0)
    dataswarm: rx: {"params":{"result":0,"state":0,"task-id":"A6AEF6F6-1C19-4DC9-9C0F-C88BC77A1F40"},"method":"task-update","jsonrpc":"2.0"}
    dataswarm: worker 127.0.0.1:45594 rx: task-update
    dataswarm: task A6AEF6F6-1C19-4DC9-9C0F-C88BC77A1F40 is active at worker
    dataswarm: rx: {"params":{"result":1,"state":1,"task-id":"A6AEF6F6-1C19-4DC9-9C0F-C88BC77A1F40"},"method":"task-update","jsonrpc":"2.0"}
    dataswarm: worker 127.0.0.1:45594 rx: task-update
    dataswarm: task A6AEF6F6-1C19-4DC9-9C0F-C88BC77A1F40 is done at worker
    dataswarm: json tx: {"result":[{"params":{"result":1,"state":1,"task-id":"A6AEF6F6-1C19-4DC9-9C0F-C88BC77A1F40"},"method":"task-update","jsonrpc":"2.0"}],"id":2,"jsonrpc":"2.0"}
    dataswarm: client disconnect (127.0.0.1:45598): Connection reset by peer
    tcp: disconnected from 127.0.0.1 port 45598
    ^C

One the worker:

    $ ./ds_worker -dall -mlocalhost -p1234
    dataswarm: checking /tmp/dataswarm-worker-1000/blob for blobs to recover...
    dataswarm: done recovering blobs
    dataswarm: 0 blobs, 0 MB inuse, 7988 MB avail, 7988 MB total
    dataswarm: checking /tmp/dataswarm-worker-1000/task for tasks to recover...
    dataswarm: done recovering tasks
    dataswarm: 0 tasks recovered using 0 MB disk
    dataswarm: inuse: 0 cores, 0 MB memory, 0 MB disk
    dns: looking up name localhost
    dns: localhost is 127.0.0.1
    tcp: connecting to 127.0.0.1 port 1234
    dataswarm: json tx: {"params":{"type":"worker"},"method":"handshake","jsonrpc":"2.0"}
    dataswarm: json tx: {"params":{"hello":"manager"},"method":"status-report","jsonrpc":"2.0"}
    dataswarm: rx: {"id":0,"params":{"state":"active","resources":{},"task-id":"A6AEF6F6-1C19-4DC9-9C0F-C88BC77A1F40","command":"exit 123"},"method":"task-submit","jsonrpc":"2.0"}
    dataswarm: task A6AEF6F6-1C19-4DC9-9C0F-C88BC77A1F40 created
    dataswarm: json tx: {"result":0,"id":0,"jsonrpc":"2.0"}
    wq: started process 17366: exit 123
    dataswarm: task A6AEF6F6-1C19-4DC9-9C0F-C88BC77A1F40 active -> active
    dataswarm: json tx: {"params":{"result":0,"state":0,"task-id":"A6AEF6F6-1C19-4DC9-9C0F-C88BC77A1F40"},"method":"task-update","jsonrpc":"2.0"}
    dataswarm: inuse: 0 cores, 0 MB memory, 0 MB disk
    dataswarm: inuse: 0 cores, 0 MB memory, 0 MB disk
    dataswarm: task A6AEF6F6-1C19-4DC9-9C0F-C88BC77A1F40 active -> done
    dataswarm: json tx: {"params":{"result":1,"state":1,"task-id":"A6AEF6F6-1C19-4DC9-9C0F-C88BC77A1F40"},"method":"task-update","jsonrpc":"2.0"}
    ds_worker_main_loop: Connection reset by peer
    tcp: disconnected from 127.0.0.1 port 1234
    ^C

On the client:

    $ ./min_client 
    DataSwarm:DEBUG:Connecting to 127.0.0.1:1234...
    DataSwarm:DEBUG:rx: {'result': 'A6AEF6F6-1C19-4DC9-9C0F-C88BC77A1F40', 'id': 1, 'jsonrpc': '2.0'}
    {'result': 'A6AEF6F6-1C19-4DC9-9C0F-C88BC77A1F40', 'id': 1, 'jsonrpc': '2.0'}
    DataSwarm:DEBUG:rx: {'result': [{'params': {'result': 1, 'state': 1, 'task-id': 'A6AEF6F6-1C19-4DC9-9C0F-C88BC77A1F40'}, 'method': 'task-update', 'jsonrpc': '2.0'}], 'id': 2, 'jsonrpc': '2.0'}
    {'result': [{'params': {'result': 1, 'state': 1, 'task-id': 'A6AEF6F6-1C19-4DC9-9C0F-C88BC77A1F40'}, 'method': 'task-update', 'jsonrpc': '2.0'}], 'id': 2, 'jsonrpc': '2.0'}
